### ISSUE TITLE
[Enhancement]: Add custom path settings

### ISSUE THREAD

**Issue by maartenbrakkee** [advplyr/audiobookshelf#385]
### Describe the feature/enhancement

For reverse proxy usage with a subfolder instead of a subdomain. E.g., `/audiobookshelf/` instead of `/`.

**Comment by unclebacon-live** [comment]
I second this.

**Comment by whiskywaffel** [comment]
Would be a nice. Is there any chance this gets implemented?

**Comment by sandreas** [comment]
+1

**Comment by sandreas** [comment]
@advplyr 

How is the state of this issue? You said in #570: 

> we need to add a base url from env variables for the client side to work. Then there is the server and mobile apps. I've attempted to add this once before and it was more complicated then it seems. 

So there is work to do in:
- Client
- Server
- Mobile Apps

I did not find a branch with your failed attempt, but I would really like to help you with this issue and submit a PR, if you don't have the time. I would just need your help to point me into the right direction... the collected parts of code above might be a first step, but I do not know, what code parts may be also critical / necessary... 

Should I fork and create a branch / PR for discussion? Or would you just list the things you tried out and failed with?

Thank you.

**Comment by advplyr** [comment]
Thanks, I could use some help.
The first step is the base url will need to be passed in as an environment variable for docker or specified in the config file for debian.

The part where I got hung up was passing that environment variable to the Nuxt.js app so that we can set the base url in the nuxt.config.js file. I'm not sure the best way to handle that. I think setting that would handle the web app client portion of this.

I'm not sure exactly what would be needed on the server yet.
The mobile apps would need to store the base url separately. I haven't gotten too deep into this since I got hung up on the web app client.

**Comment by advplyr** [comment]
Oh great, how are you passing the env variable to the nuxt app?

**Comment by sandreas** [comment]
> Oh great, how are you passing the env variable to the nuxt app?

That is something I am working at. I tried some static options first and it worked with the options `router` and `baseUrl`:

```
  router: {
    base: '/audiobooks'
  },
  axios: {
    // baseURL: process.env.serverUrl || ''
    baseURL: "/audiobooks/"
  },
```

The `nuxt`-config is very fiddly here. A missing `/` or one too much at the wrong place and it does not work at all. Seems that this is influenced by the server config in `dev.js`. 

I'm currently trying to figure out a reliable way to make this work and if I have this, I'll try the dynamic configuration. There is also this ticket: https://github.com/nuxt/nuxt.js/pull/7312 which might help to configure it runtime, so that one does not need to retranspile the whole client for changing the `baseUrl`. 

I'm pretty busy atm, maybe this takes some time.

**Comment by sandreas** [comment]
> Prefix is better since that is all we want passed into the env variable.

Ok, I'll regard that. Currently it is an experiment - not a working thing. Maybe I have to rethink everthing.

> I was having the same issue you are working on. We can use the runtime config but the problem I had was actually accessing that env from the client. The env's available on the server aren't available in the nuxt app.

I'm aware of this. I would prefer this

```
router: {
  base: ""
},
axios: {
  baseURL: ""
},
```

because  without a leading slash using relative references it would work everywhere without any runtime changes or variables. It seems that `nuxt` does not allow `` (empty string) and replaces it with `/`. I would see that as a `nuxt` bug, because it makes a very important feature unusable - REAL relative paths. That is what I am investigating.

**Comment by advplyr** [comment]
@sandreas I missed your last comment here. Are you still working on this? I can look into this today

**Comment by sandreas** [comment]
Not really, my son was born 6 days ago, now I don't have the time. Would love to see you look into this, otherwise this will take a lot of time, the next 3 weeks no development ;)

**Comment by advplyr** [comment]
@sandreas Congrats! Thanks for the update

**Comment by RichardMawdsley** [comment]
Hope your new son is doing well @sandreas! - He's what, 4/5 weeks old now? - He must be sleeping through the night and barely disturbing you, right?.. right!? :D

I have no way of helping with the development other then offering to be a guinea pig beta tester when it gets there!

Fingers crossed for this in the near future.

**Comment by sandreas** [comment]
@RichardMawdsley 

> Hope your new son is doing well @sandreas!

Yeah he is alright thank you.

> He must be sleeping through the night and barely disturbing you, right?.. right!? :D

Unfortunately this is not the way it goes ;-) However, its surprising how insomnia gets you back to development. But  @advplyr was very right: nuxt is really nasty here. I tried many things to get this working reliably but atm I'm out of ideas...  My branch contains a PARTIAL solution for ONE problem and combines multiple points of failure to one. But since I'm not a nuxt expert and it was really frustrating to debug this, I'm not going to invest time here in the next weeks. 

I have a lot of projects going on and since this problem would be "nice to have" and is not mandatory for my system to work, I deferred my efforts to get this working.

**Comment by 1nVitr0** [comment]
Am I correct in the assumption, that this is already implemented using `ROUTER_BASE_PATH`, but doesn't get picked up by nuxtjs? (Whis is because nuxt has been compiled to [static](https://github.com/advplyr/audiobookshelf/blob/dadd41cb5c221c8c79f1a26a0bebd3d740b51539/client/nuxt.config.js#L6) files)

I am happy to help here in any way possible. Just checking the desired behavior here: Do you intend to solve this inside the server code, or do you want to change the nuxt target type?

**Comment by advplyr** [comment]
> Am I correct in the assumption, that this is already implemented using `ROUTER_BASE_PATH`, but doesn't get picked up by nuxtjs? (Whis is because nuxt has been compiled to [static](https://github.com/advplyr/audiobookshelf/blob/dadd41cb5c221c8c79f1a26a0bebd3d740b51539/client/nuxt.config.js#L6) files)

Correct. The server side part should be good to go although I haven't tested that env variable in a while.

> I am happy to help here in any way possible. Just checking the desired behavior here: Do you intend to solve this inside the server code, or do you want to change the nuxt target type?

I'm not sure what the best way to solve it is. My initial thought is that we have the server build Nuxt programmatically. I haven't spent a lot of time looking at the internals of Nuxt but there is some info about this approach https://nuxtjs.org/docs/internals-glossary/nuxt/.

I would really appreciate your help or feedback if you want to dig into any of this. I'm also active on the Discord/Matrix servers if you want to chat about it.

**Comment by advplyr** [comment]
I took a non-traditional approach when first setting this up because I tend to learn as I build, so I just didn't know at the time.

**Comment by 1nVitr0** [comment]
Thats the best way to learn things, at least for me. Give a few days to fiddle aroundwith it, haven't used nuxt in a while. Given the amount of free time currently, it will probably be a bit hacky, bu I'll make sure it's robust (That is if I manage to find one).

This is an amazing project, I spend 2+ hours a day listening to audibooks. So I will definitely invest more time into this in the future.

**Comment by 1nVitr0** [comment]
Right. After a little fiddling and digging through nuxtjs changelogs, I have come to the following conclusion:

- NuxtJS does [NOT natively support](https://github.com/nuxt/nuxt/pull/8520) a dynamic base path when using a static build target
- Options to get around this are:
  1. Recompiling the target when the base path changes
  2. Running nuxt using node directly and proxying express requests to it (alternatively using nginx or similar around it)
  3. Manually updating the build files when the base path changes

I think they all have their pros and cons

- `i.` is robust and includes minimal code changes, but bloats the docker image with nuxt dependencies and drastically slows down startup
- `ii.` is robust, efficient and doesn't bloat the docker image very much, as nuxtjs can compile it's dependencies into the build. But it requires some drastic code changes, as we can no longer rely on the static serving of files to detect the request's targets
- `iii.` requires minimal code changes, is very efficient and does minimally startup performance. But it relies on nuxt compiling into a known structure and may break when updating to a new version

I am currently dabbling with `iii.` as a temporary solution. That does seem to work until the login screen, after which my version currently crashes. (It does not connect to the websocket, so that may be the issue). For a permanent solution, I think either `i.` or `ii.` are fine, although I personally lean towards `ii.` as `node` is used for the rest of the application anyways.

**Comment by Danathar** [comment]
I'd like to present an additional use case, which is how I use it although it seems that everyone already supports the idea. I currently use Nginx's reverse proxy with different subfolder paths to distinguish between Apache Guacamole (a remote desktop app over the web that also utilizes websockets) and Wetty (a web-based terminal). I would prefer to use a subfolder URL for Audiobookshelf, as it would eliminate the need to obtain another certificate for an additional subdomain. I use Let's Encrypt for my certificates which makes the process of obtaining and managing certificates relatively simple and automated. However, having fewer certificates to manage would still simplify the process.

**Comment by sandreas** [comment]
>  I use Let's Encrypt for my certificates which makes the process of obtaining and managing certificates relatively simple and automated. However, having fewer certificates to manage would still simplify the process.

A little off-topic: You could use letsencrypt wildcard certificates (`*.domain.com`), which work perfectly fine on my system.

**Comment by Danathar** [comment]
> > I use Let's Encrypt for my certificates which makes the process of obtaining and managing certificates relatively simple and automated. However, having fewer certificates to manage would still simplify the process.
> 
> A little off-topic: You could use letsencrypt wildcard certificates (`*.domain.com`), which work perfectly fine on my system.

Yea I guess it’s slightly off topic. I forgot to use wildcard certs when I first setup my system. I don’t think it’s trivial to convert once you have gone down the road of individual certs. I think I would need to revoke all my certs and start over.

**Comment by MagiX13** [comment]
I could be entirely wrong here as I never worked with anything from the underlying stack, but isn't this supported in nuxt using the runtime config, as described with basically the exact example [here](https://axios.nuxtjs.org/options/#runtime-config)
```
publicRuntimeConfig: {
    axios: {
      browserBaseURL: process.env.BROWSER_BASE_URL
    }
  },
privateRuntimeConfig: {
    axios: {
      baseURL: process.env.BASE_URL
    }
  }
```
Giving a bit more context on the runtime config from [here](https://v2.nuxt.com/docs/directory-structure/nuxt-config#runtimeconfig)

>The runtime config has built-in [dotenv ](https://github.com/motdotla/dotenv)support for better security and faster development. **The runtime config is added to the Nuxt payload so there is no need to rebuild in order to update the runtime configuration**...

**Comment by advplyr** [comment]
@MagiX13 That is for axios and not for nuxt which is generating all the paths during build. There is a detailed discussion in #1746

**Comment by gardiol** [comment]
Is there any news on this? I would also like this feature if not yet implemented.

**Comment by gaarysn** [comment]
What shoud I do if my server already have subdomain (like srv.server.com). How to configure nginx and ssl to audiobookshelf.srv.server.com?

**Comment by Danathar** [comment]
> Hi. What shoud I do if my server already have subdomain (like srv.server.com). How to configure nginx and ssl to audiobookshelf.srv.server.com? Will audiobookshelf work like this?

Create a server block in nginx like:

```
server {
    listen 80;
    server_name audiobookshelf.srv.server.com;

    location / {
        proxy_pass http://localhost:13378; # Replace with the port Audiobookshelf is running on
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
}
```

You could set that listen to whatever you like, but if you use 443 you'll need to set up certs. I set mine up with letsenrypt. The block above was spit out by AI, but it's similar to mine except I've added encryption.

**Comment by farfalleflickan** [comment]
@1nVitr0 I do not mean to nag, but would it be possible to know the progress status on this?

**Comment by mikiher** [comment]
I added a small writeup in discussion #3535 to lay out my plan for supporting this within the current codebase.

I already have this mostly working on my dev environement, but I wanted to solicit some feedback before I continue, since the changes I'm suggesting are somewhat subtle.

**Comment by 4ch1m** [comment]
After reading all the comments it's pretty obvious that this issue is a hard nut to crack. :smile: 

However, this also seems to be the [most requested](https://github.com/advplyr/audiobookshelf/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) feature at the moment.

And although partial progress has been reported, plus @mikiher already compiled a very helpful [gist](https://github.com/advplyr/audiobookshelf/discussions/3535)... I'm still confused. :exploding_head:

So... is anyone still actively working on this? :thinking: 

I don't want to be impatient or demanding.
I'm just asking to avoid that someone interested (and maybe skilled) is hesitating to tackle the issue based on the false assumption that someone else is already half-way there.

Thanks!

**Comment by mikiher** [comment]
I am working on it, and it is indeed a bit of a hard nut to crack, as you say (as you may have seen there was some additional discussion in #3535 which complicated things a bit more).

I was side-tracked a bit by some other crash-causing issue I need to deal with first. 

I'll definitely get back to this one once I'm done (this isn't meant to ward off anyone who may want to tackle this on their own). I estimate it will take me an additional couple of weeks to complete.

**Comment by 4ch1m** [comment]
Thanks for bringing us up to speed, @mikiher.
That's great to hear and much appreciated. :bow: 

I would have tried my luck if no one else was working on it.
But I guess it would take me more than just a couple of weeks. :smirk: 

Let me know if you need support for testing your work/implementation.

**Comment by iamckun** [comment]
this nice，What's the latest

**Comment by mikiher** [comment]
I submitted PR #3810 to enable subdirectory support. 

Once this makes it to the server release, subdirectory access will be enabled on both server and web client.

Subdirectory support for the mobile app will also be enabled on the next mobile app release (audiobookshelf-app PR #1417).

**Comment by iamckun** [comment]
/textures/wood_default.jpg 404

**Comment by mikiher** [comment]
@iamckun Thanks for reporting. Just sent a fix for review.

**Comment by mikiher** [comment]
A few notes for on subdirectory support, for those who have not read #3535:
- Due to technical reasons (explained in #3535), you cannot change the subdirectory name (unless you override it in the code). It is hardcoded to be `/audiobookshelf`
- As of v2.18.0, the server supports both legacy (no-subdirectory) and `/audiobookshelf` access. 
- As of v2.18.0, the web client will always use `/audiobookshelf`
- Other existing clients should work seamlessly with server v2.18.0 using legacy (no-subdirectory) access.
- As of v0.9.78, the mobile app also supports `/audiobookshelf` access. To connect to a server using `/audiobookshelf`, Add a new server with address `https://<your-server-domain-and-port>/audiobookshelf`

Please let me know here if you encounter any issues with this.
